[openmp][wasm] Allow compiling OpenMP to WebAssembly

abrown opened the pull request:
This change allows building the static OpenMP runtime, `libomp.a`, as WebAssembly. It builds on the work done in [D142593] but goes further in several ways:
 - it makes the OpenMP CMake files more WebAssembly-aware
 - it conditions much more code (or code that had been refactored since [D142593]) for `KMP_ARCH_WASM` and `KMP_OS_WASI`
 - it fixes a Clang crash due to unimplemented common symbols in WebAssembly

The commit messages have more details. Please understand this PR as a start, not the completed work, for WebAssembly support in OpenMP. Getting the tests running somehow would be a good next step, e.g.; but what is contained here works, at least with recent versions of [wasi-sdk] and engines that support [wasi-threads]. I suspect the same is true for Emscripten and browsers, but I have not tested that workflow.

[D142593]: https://reviews.llvm.org/D142593
[wasi-sdk]: https://github.com/WebAssembly/wasi-sdk
[wasi-threads]: https://github.com/WebAssembly/wasi-threads

abrown committed:
[openmp][wasm] Use external symbol instead of common symbol for lock

The `gomp_critical_user` lock is typically compiled using a common
symbol, but the WebAssembly backend does not have these implemented.
This change uses an external symbol instead, which avoids compiler
crashes when compiling `#pragma omp critical` code to WebAssembly.

abrown committed:
[openmp][wasm] Inform CMake of a new WebAssembly target

This change touches CMake files only. It sets up a new `LIBOMP_ARCH`
option--`wasm32`--and provides a handy `WASM` shortcut variable for
internal CMake use like other targets do. This also disables shared
library support (WebAssembly expects a static one) and the
`LIBOMPTARGET` functionality when targeting WebAssembly.

abrown committed:
[openmp][wasm] Conditionally compile many features in WebAssembly

This change sets up definitions for the many preprocessor conditions
used in OpenMP and compiles out features that will not be supported in a
WebAssembly target. The `KMP_ARCH_WASM` target is tied to `wasm32` and
ensures 32-bit pointers are used. The `KMP_OS_WASI` target is more
tricky: it attempts to cover the subset of functionality that is
compilable by both Emscripten and wasi-sdk. Since Emscripten currently
has support for more of libc than wasi-sdk, the "WASI" name is used to
indicate that only the wasi-sdk subset is used here.

This change follows along the same lines as [D142593], though it takes
things a bit further, compiles with multiple toolchains (Emscripten
_and_ wasi-sdk), handles some new cases, and avoids for now the
`microtask_*` changes from that PR. Many thanks to @arsnyder16 for
providing some initial indication that this all was possible!

[D142593]: https://reviews.llvm.org/D142593

Co-authored-by: Atanas Atanasov <atanas.atanasov@intel.com>

shiltian reviewed (commented):
We will also want (have) to disable certain features such as OMPD and OMPT for WASM.

shiltian commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -4746,10 +4746,13 @@ OpenMPIRBuilder::getOrCreateInternalVariable(Type *Ty, const StringRef &Name,
-        M, Ty, /*IsConstant=*/false, GlobalValue::CommonLinkage,
-        Constant::getNullValue(Ty), Elem.first(),
-        /*InsertBefore=*/nullptr, GlobalValue::NotThreadLocal, AddressSpace);
+    auto linkage = this->M.getTargetTriple().rfind("wasm32") == 0
```
I'd make it a separate patch for this change

jdoerfert commented:
> This change allows building the static OpenMP runtime, libomp.a, as WebAssembly.
If you are planning to use LTO across the user code and runtime we need to disable certain parts of OpenMP-Opt or bad things might happen.

abrown commented:
Ok, I rebased this on `main`, which fixes the conflicts due to a recent addition of the s390x target. I'm open to splitting this up into separate PRs but I'll wait until it's more clear who the primary reviewer is and what other feedback I should address.

jpeyton52 commented on openmp/runtime/src/z_Linux_asm.S:
```diff
@@ -2286,7 +2286,27 @@ KMP_PREFIX_UNDERSCORE(__kmp_unnamed_critical_addr):
+__kmp_unnamed_critical_addr:
+    .4byte .gomp_critical_user_
+    .size __kmp_unnamed_critical_addr, 4
+#endif
```
To elaborate here a bit, when a parallel region is called, the task corresponding to the parallel region (microtask_t) that is generated by the compiler is sent through:

 `__kmpc_fork_call()` -> `__kmp_fork_call()` -> `__kmp_invoke_microtask(microtask_t, gtid, tid, argc, ...)` -> call the microtask with `argc` number of "shareds"

where the varargs component is a list of pointers to shared variables (or firstprivates). There is a limited, platform-independent implementation which WASM is using inside `z_linux_util.cpp` at around line 2492 which can handle up to 15 args. Anymore than that, and the runtime will error out. The test in `openmp/runtime/test/misc_bugs/many-microtask-args.c` is an example of this limitation.

In `z_Linux_asm.S` there are platform-dependent assembly versions of this function. For WASM, it would be best if an assembly version of the `__kmp_invoke_microtask()` function is created to handle any number of "shareds" passed in.

abrown replied:
> For WASM, it would be best if an assembly version of the __kmp_invoke_microtask() function is created to handle any number of "shareds" passed in.

I'm not sure if that is even possible here. WebAssembly does not have variable arguments, so to call the microtask I suspect we'll always need a `switch` to dispatch to the right signature.

jpeyton52 replied:
Ah OK, I understand the limitation now.

abrown commented:
I updated this PR with @jpeyton52's suggestions and rebased again on the latest `main` branch. The only code-related discussion I see that is still needed is about `__kmp_invoke_microtask` &mdash; must that be figured out for this PR or can that be a follow-on improvement (if indeed it is possible to do)?

jpeyton52 reviewed (approved):
Thanks for the patch!

As the changes are now. I believe they are good enough as an initial support patch for WebAssembly.

arsnyder16 commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -4982,10 +4982,13 @@ OpenMPIRBuilder::getOrCreateInternalVariable(Type *Ty, const StringRef &Name,
-        /*InsertBefore=*/nullptr, GlobalValue::NotThreadLocal, AddressSpace);
+    auto Linkage = this->M.getTargetTriple().rfind("wasm32") == 0
+                       ? GlobalValue::ExternalLinkage
+                       : GlobalValue::CommonLinkage;
```
@abrown Hey Andrew, this seems to break linking under wasm.

Previously when it externally linked it was required to define these variables through assembly, hence you updates to z_Linux_asm.S specially `.global .gomp_critical_user_.reduction.var` are actually no longer required. 

But now that they are common linkage they suffer from duplicate symbol problem.

For critical sections this can avoided if you manually add a name to the critical section that does not conflict 
`omp critical(UniqueValue)`

but for reduction allowing them to be named is not an option, so if multiple compilation units use reduction you will introduce a duplicate symbols

```
wasm-ld: error: duplicate symbol: .gomp_critical_user_.reduction.var
>>> defined in lib/libTest.a(a.cpp.o)
>>> defined in lib/libTest.a(b.cpp.o)
```

I am wonder if something like `InternalLinkage` is more appropriate

arsnyder16 commented on openmp/runtime/src/kmp_platform.h:
```diff
@@ -76,14 +77,20 @@
 #define KMP_OS_SOLARIS 1
 #endif
 
+#if (defined __wasi__) || (defined __EMSCRIPTEN__)
```
@abrown 
This is conflating wasi and the emscripten toolchain. They are two different cases where emscripten mimics linux/posix system by statically linking musl into the wasm file.

The changes in this PR will not allow openmp to compile under emscripten, the changes here will
https://github.com/llvm/llvm-project/pull/95169

abrown replied:
Good finds, both of them. Let's discuss in #95169.
